src/lowering/desugar.ml

```diff
@@ -198,6 +198,9 @@ and exp' at note = function
   | S.WhileE (e1, e2) -> (whileE (exp e1) (exp e2)).it
   | S.LoopE (e1, None) -> I.LoopE (exp e1)
   | S.LoopE (e1, Some e2) -> (loopWhileE (exp e1) (exp e2)).it
+  | S.(ForE (p, ({it=CallE (({it=DotE (arr, proj); _} as c0), c1, c2); _} as e1), e2))
+      when T.is_array arr.note.S.note_typ && proj.it = "vals"
+    -> (exp (rewrite_for_to_while p arr proj c0 c1 c2 e1 e2)).it
```

Comment:
Why call exp on the result of rewrite_for_to_while, which requires you to generate Source and not just directly call rewrite_for_to_while and have it generate the IR using the helpers in Construct.ml (plus any more you want to add).

Reply:
(in the style of S.While (e1, e2) -> (whileE (exp e1) (exp e2)).it above).

The final .it project is ugly but only a small price to pay.

Reply:
Done.